### PR TITLE
ascanrulesBeta: Fix Source Code Disclosure File Inc. False Positives

### PR DESCRIPTION
**Pull request by kingthorin**
## Overview
- CHANGELOG > Added fix note.
- SourceCodeDisclosureFileInclusionScanRule > Adjusted behavior.
- SourceCodeDisclosureFileInclusionScanRule > Added tests to assert the expected behaviors.
- ascanrulesBeta.html > Added details on the error conditions skip.

## Related Issues
- Fixes zaproxy/zaproxy#8517

## Checklist
- [na] Update help
- [x] Update changelog
- [x] Run `./gradlew spotlessApply` for code formatting
- [x] Write tests
- [x] Check code coverage (Now 68.8%, was 38.7%)
- [x] Sign-off commits
- [x] Squash commits
- [x] Use a descriptive title

### CODE REVIEW

File: addOns/ascanrulesBeta/CHANGELOG.md [hunk 1]
```diff
@@ -4,15 +4,16 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
-
+### Fixed
+- Potential false positives in the Source Code Disclosure - File Inclusion scan rule (Issue 8517).
```
Comment (thc202) [inline addOns/ascanrulesBeta/CHANGELOG.md:8]: Would be good if it was more clear what FP is this addressing, for the record.

File: addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java [hunk 2]
```diff
@@ -235,10 +235,13 @@ public void scan(HttpMessage originalmsg, String paramname, String paramvalue) {
                     DiceMatcher.getMatchPercentage(
                             originalmsg.getResponseBody().toString(),
                             randomfileattackmsg.getResponseBody().toString());
-            if (originalversusrandommatchpercentage > this.thresholdPercentage) {
-                // the output for the "random" file does not sufficiently differ. bale out.
+            if (randomfileattackmsg.getResponseBody().length() == 0
+                    || originalversusrandommatchpercentage > this.thresholdPercentage) {
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java:239]: Extract a method.

File: addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java [hunk 3]
```diff
@@ -235,10 +235,13 @@ public void scan(HttpMessage originalmsg, String paramname, String paramvalue) {
...
+                // the output for the "random" file does not sufficiently differ or was empty. bale
+                // out.
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java:241]: Remove it instead.

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java [hunk 4]
```diff
@@ -85,4 +101,141 @@ void shouldReturnExpectedExampleAlert() {
...
         assertThat(alert.getConfidence(), is(equalTo(Alert.CONFIDENCE_MEDIUM)));
     }
+
+    @Test
+    void shouldSkipUrlParams() {
+        // Given
+        HttpMessage msg = createMessage("/test/");
+        NameValuePair nvp = new NameValuePair(NameValuePair.TYPE_URL_PATH, "inc", "", 0);
+        // When
+        rule.scan(msg, nvp);
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:111]: There's no reason to call this one directly.

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java [hunk 5]
```diff
@@ -85,4 +101,141 @@ void shouldReturnExpectedExampleAlert() {
...
+    void shouldNotContinueIfRandomBodyIsEmpty() throws IOException {
+        // Given
+        String path = "/shouldNotContinueIfRandomBodyIsEmpty";
+        nano.addHandler(new FiHandler(path, Response.Status.OK, ""));
+        HttpMessage msg = this.getHttpMessage(path + "?inc=bar");
+        rule.init(msg, parent);
+        // When
+        rule.scan();
+        // Then
+        assertThat(httpMessagesSent.size(), is(equalTo(2)));
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:138]: Not sure asserting exact number of requests is good here, if the rule is improved and it now sends more/less we have to go through all of them again, if you want to check that the rule is actually sending something check it's not empty.

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java [hunk 6]
```diff
@@ -85,4 +101,141 @@ void shouldReturnExpectedExampleAlert() {
...
+        HttpMessage m = httpMessagesSent.get(1);
+        assertThat(m.getResponseBody().toString(), is(equalTo("")));
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:140]: Should also assert no alerts (same for following one).

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java [hunk 7]
```diff
@@ -85,4 +101,141 @@ void shouldReturnExpectedExampleAlert() {
...
+        // When
+        rule.scan();
+        // Then
+        assertThat(httpMessagesSent.size(), is(equalTo(2)));
+        HttpMessage m = httpMessagesSent.get(1);
+        assertThat(m.getResponseBody().toString(), is(equalTo(DEFAULT_RESPONSE_STRING)));
+    }
+
+    @Test
+    void shouldAlertWhenConditionsRightAndPhpExtensionMatchesContent() throws IOException {
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:159]: Should be paremeterized or renamed to be generic (same for the following).

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java (hunk 4 above)
Comment (kingthorin) [reply addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:111]: So get rid of this test?

Comment (thc202) [reply addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:111]: It should call init/scan.

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java (hunk 7 above)
Comment (kingthorin) [reply addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:159]: What would be parameterized in the next one, no extension is a single case

File: addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java [hunk 8]
```diff
@@ -206,6 +206,9 @@ public void scan(HttpMessage msg, NameValuePair originalParam) {
      */
     @Override
     public void scan(HttpMessage originalmsg, String paramname, String paramvalue) {
+        if (isClientError(getBaseMsg()) || isServerError(getBaseMsg())) {
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java:209]: Worth updating the help.

File: addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java [hunk 9]
```diff
@@ -434,6 +437,12 @@ public void scan(HttpMessage originalmsg, String paramname, String paramvalue) {
         }
     }
 
+    boolean isEmptyOrTooSimilar(
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java:440]: `private`

File: addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java [hunk 10]
```diff
@@ -434,6 +437,12 @@ public void scan(HttpMessage originalmsg, String paramname, String paramvalue) {
...
+            HttpMessage randomfileattackmsg, int originalversusrandommatchpercentage) {
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/main/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRule.java:441]: `HttpMessage msg, int matchPercentage`

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java [hunk 11]
```diff
@@ -85,4 +103,161 @@ void shouldReturnExpectedExampleAlert() {
         assertThat(alert.getRisk(), is(equalTo(Alert.RISK_HIGH)));
         assertThat(alert.getConfidence(), is(equalTo(Alert.CONFIDENCE_MEDIUM)));
     }
+
+    @Test
+    void shouldSkipUrlParams() {
+        // Given
+        HttpMessage msg = createMessage("/test/?inc=foo");
```
Comment (thc202) [inline addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java:110]: I'd not expect the query parameter.

**Comment by kingthorin** [comment]
Done

File: addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java (hunk 11 above)
Comment (thc202) [reply addOns/ascanrulesBeta/src/test/java/org/zaproxy/zap/extension/ascanrulesBeta/SourceCodeDisclosureFileInclusionScanRuleUnitTest.java]: Needs `scannerParam.setTargetParamsInjectable(ScannerParam.TARGET_URLPATH);`.

**Comment by kingthorin** [comment]
Done

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**